### PR TITLE
Create index.d.ts for Typescript

### PR DESCRIPTION
**Pull request by yoonhoGo**

### CODE REVIEW

**Comment by lorensr** [comment]
Thanks! Wondering how to test this (eg w/ VSCode) and publish to npm?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  import { DataSource, DataSourceConfig } from 'apollo-datasource'
+  import * as mongoose from 'mongoose'
+
+  export {
+    DataSourceConfig
+  }
+
+  interface Options { ttl: number }
+
+  export abstract class MongoDataSource<T extends mongoose.Document, TContext = any> extends DataSource {
```
Comment (9at8) [inline index.d.ts]:
```suggestion
  export abstract class MongoDataSource<T extends mongoose.Document, TContext = any> extends DataSource<TContext> {
```

`DataSource` is also a generic type, which optionally takes in the context type, which is what we should do imo.

Comment (9at8) [reply index.d.ts]: Also, I don't think the `collection` type makes these typings compatible with the native mongodb connector, which has a generic type for collection

**Comment by 9at8** [comment]
@lorensr [this](https://cameronnokes.com/blog/the-30-second-guide-to-publishing-a-typescript-package-to-npm/) explains in depth what to do.

*TLDR;* Add a `"types": "index.d.ts"` to `package.json` before publishing

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    collection: mongoose.Collection
+    model?: mongoose.Model<T>
+    constructor(collection: mongoose.Collection | mongoose.Model<T>)
+    public initialize(config: DataSourceConfig<TContext>): void
+    protected findOneById(id: string, { ttl }?: Options): T | null
+    protected findManyByIds(ids: string[], { ttl }?: Options): (T| null)[]
```
Comment (9at8) [inline index.d.ts]: The ids are supposed to be `ObjectId`s, at least for the native driver. Can you confirm @lorensr

Comment (lorensr) [reply index.d.ts]: Yes. And if it comes in as a string argument, I use this function to convert:

image

**Comment by gayansanjeewa** [comment]
Hi guys, 

Any progress on this feature..? We'd love to use this in one of our projects and for that we need the Typescript support.

Thanks in advance. 
Gayan

**Comment by lorensr** [comment]
@gayansanjeewa we're waiting on either @yoonhoGo to make the suggested changes and someone to test it, or for someone to create another PR with the changes

**Comment by lorensr** [comment]
Merged #13, released in `0.2.4`
